proposals/0000-add-sequence-based-init-and-merge-to-dictionary.md

```diff
@@ -0,0 +1,185 @@
+
+    init<S: SequenceType where S.Generator.Element == (Key, Value)>(
+        merging sequence: S, 
+        @noescape combine: (Value, Value) throws -> Value
+        ) rethrows
+
+This example shows how one could keep the first value of all those supplied for a duplicate key:
+
+    let letterDict2 = Dictionary(merging: duplicateLetters, combine: { (first, _) in first })
+    // ["b": 2, "a": 1]
+
+Or the largest value for any duplicate keys:
+
+    let letterDict3 = Dictionary(merging: duplicateLetters, combine: max)
+    // ["b": 4, "a": 3]
+
+At other times the merging initializer could be used to intentionally combine values for duplicate keys. Donnacha Oisín Kidney wrote a neat `frequencies()` method for sequences as an example of such a use in the thread:
+
+    extension SequenceType where Generator.Element: Hashable {
+        func frequencies() -> [Generator.Element: Int] {
+            return Dictionary(merging: self.lazy.map { v in (v, 1) }, combine: +)
+        }
+    }
+    [1, 2, 2, 3, 1, 2, 4, 5, 3, 2, 3, 1].frequencies()
+    // [2: 4, 4: 1, 5: 1, 3: 3, 1: 3]
+
+This proposal also includes new mutating and non-mutating methods for `Dictionary` that merge the contents of a sequence of `(Key, Value)` tuples into an existing dictionary:
+
+    mutating func mergeContentsOf<S: SequenceType where S.Generator.Element == (Key, Value)>(_ sequence: S, @noescape combine: (Value, Value) throws -> Value) rethrows
+    mutating func mergedWith<S: SequenceType where S.Generator.Element == (Key, Value)>(_ sequence: S, @noescape combine: (Value, Value) throws -> Value) rethrows -> [Key: Value]
+
+As above, there are a wide variety of uses for the merge. The most common might be merging two dictionaries together:
+
+    // Adding default values
+    let defaults: [String: Bool] = ["foo": false, "bar": false, "baz": false]
+    var options: [String: Bool] = ["foo": true, "bar": false]
+    options.mergeContentsOf(defaults) { (old, _) in old }
+    // options is now ["foo": true, "bar": false, "baz": false]
+    
+    // Summing counts repeatedly
+    var bugCounts: [String: Int] = ["bees": 9, "ants": 112, ...]
+    while bugCountingSource.hasMoreData() {
+        bugCounts.mergeContentsOf(bugCountingSource.countMoreBugs(), combine: +)
+    }
+
+    
+## Detailed design
+
+The design is simple enough -- loop through the sequence and update the new or existing dictionary. As an optimization, it makes sense to push the merging down to the variant storage layer to avoid having to do duplicate hash/index lookups when duplicate keys are found.
+
+Collected in one place, the new APIs for `Dictionary` look like this:
+
+```swift
+init?<S: SequenceType where S.Generator.Element == (Key, Value)>(
```

Comment:
These look good, thank you!

I would recommend to write doc comments for them right in the proposal.